specs/agents/tracing-spans-compress.md

```diff
@@ -0,0 +1,199 @@
+      The minimum count is 2 as a composite span represents at least two spans.
+    - `end`: The end timestamp of the last compressed span.
+      The net duration of all compressed spans is equal to the composite spans' `duration`.
+      The gross duration (including "whitespace" between the spans) is equal to `compressed.end - timestamp`.
+
+#### Turning compressed spans into a composite span
+
+Spans have a `compress` method.
+The first time `compress` is called on a regular span, it becomes a composite span.
+
+```java
+void compress(Span other) {
+    if (compressed.count == 0) {
+        compressed.count = 2
+    } else {
+        compressedCount++
+    }
+    endTimestamp = max(endTimestamp, other.endTimestamp)
+}
+```
+
+#### Effects on metric processing
+
+As laid out in the [span destination spec](tracing-spans-destination.md#contextdestinationserviceresource),
+APM Server tracks span destination metrics.
+To avoid compressed spans to skew latency metrics and cause throughput metrics to be under-counted,
+APM Server will take `composite.count` into account when tracking span destination metrics.
+
+### Compression algorithm
+
+#### Eligibility for compression
+
+A span is eligible for compression if all the following conditions are met
+- It's an exit span
+- The trace context of this span has not been propagated to a downstream service
+
+The latter condition is important so that we don't remove (compress) a span that may be the parent of a downstream service.
+This would orphan the sub-graph started by the downstream service and cause it to not appear in the waterfall view.
+
+```java
+boolean isCompressionEligible() {
+    return exit && !context.hasPropagated
+}
+```
+
+#### Span buffering
+
+Non-compression-eligible spans may be reported immediately after they have ended.
+When a compression-eligible span ends, it does not immediately get reported.
+Instead, the span is buffered within its parent.
+A span/transaction can buffer at most one child span.
+
+Span buffering allows to "look back" one span when determining whether a given span should be compressed.
+
+A buffered span gets reported when
+1. its parent ends
+2. a non-compressible sibling ends
+
+```java
+void onSpanEnd() {
+    if (isCompressionEligible()) {
+        if (parent.hasBufferedSpan()) {
+            parent.tryCompress(this)
+        } else {
+            parent.buffered = this
+        }
+    } else { 
+        report(buffered)
+        report(this)
+    }
+}
+```
+
+#### Compression
+
+On span end, we compare each [compression-eligible](tracing-spans-compress.md#eligibility-for-compression) span to it's previous sibling.
+
+If the spans are of the same kind but have different span names,
+we compress them using the [Consecutive-Same-Kind compression strategy](tracing-spans-compress.md#consecutive-same-kind-compression-strategy).
+
+If the spans are of the same kind, and have the same name,
```

Comment:
'same kind' --> let's specify this here again! --> which fields need to be equal?

Reply:
I'd rather link to the location where that's defined than repeating it.